Improved database wrapper to get conversations, save data and track status (issue #6)

zezere opened the pull request:
See the diagram on this Miro board:

https://miro.com/welcomeonboard/Zm5sYXI3cy8vd3hZWWF5UXhKamI0aStXK2NOMGdtU3NsQjJxaUR1QmZ4V1ZXRS9Md0lWb3R4RWwyVXRpNkRUZXcrVkFzVVZ2OGlqYVVVT3UvWURQR3ZHQ0dQekJVWDJtRFdObkdoZGVMd2JBL3ZIR0xoRnJGOXdsNWdBa3RtSGlzVXVvMm53MW9OWFg5bkJoVXZxdFhRPT0hdjE=?share_link_id=171781479527

zezere committed:
Improved methods to get conversations

zezere committed:
Added remaining data update functions

zezere committed:
Solved some technical debt

yellowdolphin commented:
Code review sucks on github, I only get a little popup window and all my input just disappeared. So I do it as a comment.

## ConversationsDB
- Can we have `db_name` and `bot_emails` with defaults in the signature of __init__ rather than hard-coded in the function body?
- move DatabaseManager import up to other module imports

## update_data_after_step2
- update return value (bool) in type hints

## get_scheduled_conversations
- discuss: add schedule datetime and num_reminders_sent

yellowdolphin reviewed (changes requested):
See comment https://github.com/zezere/email-bot/pull/12#issuecomment-2838689416

zezere committed:
Fixes as requested in PR review

zezere commented:
Made changes, updated this pull request, @yellowdolphin you can review again.
Added some answers and comments to the [Miro Board](https://miro.com/welcomeonboard/Zm5sYXI3cy8vd3hZWWF5UXhKamI0aStXK2NOMGdtU3NsQjJxaUR1QmZ4V1ZXRS9Md0lWb3R4RWwyVXRpNkRUZXcrVkFzVVZ2OGlqYVVVT3UvWURQR3ZHQ0dQekJVWDJtRFdObkdoZGVMd2I3M0ZZWG10WldYK3diV0xHYmtEN2p3VHhHVHd5UWtSM1BidUtUYmxycDRnPT0hdjE=?share_link_id=325766241475) as well.

In addition to requested changes, I added function to insert test data if you're creating a new empty database.

data files should not be on GitHub, because I did use a couple real email addresses in that data, so I am only commiting an empty folder "data/", and the contents of that folder I'm zipping up and uploading to our google drive, here's the link:
https://drive.google.com/file/d/1sKJzrUdbfrN7Byi2CKR0kGhEnMNvGNpp/view?usp=drive_link
I have been testing with this data.

yellowdolphin commented:
Awesome, I also finished refactoring main/bot and the dependent modules llm_handler, scheduling, utils.

If you want to use folder "data" for testing but not include/use it in the repository, let's better keep it in ".gitignore". Because you don't want to commit (by accident) any files you put in there. But if you want to use it as the place where the database files go (in production), an empty folder is ok, and we should not use it for testing then. Let me know, because currently I'm using it for testing.

I would suggest, when we have data that we want to use permanently for testing, it should go in a new folder "test" or "test/data" together with the tests depending on that data. It should be more targeted and minimal test data than the more "random" data we use now for debugging.

A little warning: the "Changes reviewed" box stays at the bottom, even if it is older than our comments.

zezere commented:
Yes I keep data in .gitignore.
But I agree that it's better to have a separate "test/" folder

yellowdolphin commented:
Questions:
- I assume, the default db_name is the production one? Otherwise, if you want to set it in your shell script and pass it to main(), we would need to decide how to do this. 
- As `check_db_status` returns `all_processes_completed`, it is trivial for main/bot to check it. But we need to decide for the MTP, how we want to handle this: should the `main` return if `all_processes_completed` is False? Do you then want to complete the processes manually? Or call `main` with an arg "restart" to ignore it and run anyway?

main looks like this now:
```python
from core.conversations_db import ConversationsDB
from bot import Bot

RESTART = False


def main():
    conv_db = ConversationsDB()

    if not conv_db.all_replies_sent():
        print("Not all replies sent yet, returning.")
        return

    all_processes_completed = conv_db.check_db_status()
    if not all_processes_completed:
        if RESTART:
            print("Not all processes completed, calling bot anyway.")
        else:
            print("Not all processes completed, returning.")
            return

    bot = Bot(conv_db)
    bot.analyze_conversations()  # step 1: set schedules & identify running conversations
    any_errors = bot.manage_running_conversations()  # step 2: write responses
    if any_errors:
        print("Failed to generate or save responses for some conversations, "
              "skipping step 3 (manage_reminders).")
        return
    bot.manage_reminders()  # step 3: process schedules & (step 4): write reminders


if __name__ == "__main__":
    main()
```

zezere commented:
no, db_name default value is test database. We will make production database when we can fetch real mails. Until then we only have test emails, so therefore we have only test database.
Not sure what you mean by the shell script.

As for what happens when not all processes completed - the email flags will be inconsistent then as well. But how to fix that, right now we don't know because we don't know what will the problem be. Let's connect everything so that it works and see what problems we have then.

However it is true that if there are incomplete processes, then ignoring it and running the program again will most likely not work and produce some wrong results, due to inconsitent flags. What I can do is add a function to reset all flags - all "analyzed", "processed", and "needs_reply" - to zeroes, and also delete everything from the process table (ps_list). Then at least you would be able to run the next loop and maybe next loop works okay.

yellowdolphin commented:
Let's discuss that in the next meeting, honestly, I don't feel comfortable with that. These flags should always be consistent by design if they are not, I would consider this a bug. Surely, there will be bugs and therefore we might want to analyze the database after a random crash, but at testing time, not in production. We should plan ahead at least towards convenient testing.

yellowdolphin commented:
Another question: What is the difference between the return value of  the ConversationsDB methods `check_db_status` and `get_unanalyzed_conversations`?
The first returns True if all processes completed.
The second returns False if track = True and at least 1 conversation has incomplete process in ps_list
Is this is not redundant? Should the bot behave any differently?

zezere commented:
I think the mistake we made is to dive into development without defining the minimum list of features that are needed for the MTP. I will define it now: for MTP we need to be able to handle one first email from a new user - and maybe a second email. This would already enable us to do first user tests. These would be controlled user tests, meaning performed with UX person present (Petra probably), so they would last 15 minutes to max 1hr each, and only one at a time. No multiple users. No replies/schedules a day later. Just one interaction with one user within the limits of max 1 hour. The purpose is to see if they understand what this thing is, how to interact with it, to see what their expectations are, and what are their feelings after that first interaction.

Let's talk during the weekend. The goal would be to decide where we draw the line to improvements. Do you have any time preference?

yellowdolphin commented:
Yes, let's talk on the weekend because now I'm totally confused. Let's change the medium.

zezere commented:
> Another question: What is the difference between the return value of the ConversationsDB methods `check_db_status` and `get_unanalyzed_conversations`? The first returns True if all processes completed. The second returns False if track = True and at least 1 conversation has incomplete process in ps_list Is this is not redundant? Should the bot behave any differently?

Function get_unanalyzed_conversation usees internal function _start_tracking, and this function only makes sense if the conversations that it will start tracking, are not tracked yet.
So the difference is that "check_db_status" checks all processes, while this function checks only processes for conversations that are being returned.

Redundancy in safety checks is good to have.
